cpp/oneapi/dal/algo/basic_statistics/backend/gpu/finalize_compute_kernel_dense_impl_dpc.cpp

```diff
@@ -0,0 +1,246 @@
+                                                               const input_t& input) {
+    result_t res;
+    auto local_desc = get_desc_to_compute<Float>(desc);
+    const auto res_op_partial = local_desc.get_result_options();
+    auto column_count = 0;
+
+    if (res_op_partial.test(result_options::min)) {
+        column_count = input.get_partial_min().get_column_count();
+    }
+    if (res_op_partial.test(result_options::sum)) {
+        column_count = input.get_partial_sum().get_column_count();
+    }
+
+    ONEDAL_ASSERT(column_count > 0);
+
+    const auto res_op = desc.get_result_options();
+    res.set_result_options(desc.get_result_options());
+
+    const auto nobs_nd = pr::table2ndarray_1d<Float>(q, input.get_partial_n_rows());
+
+    auto rows_count_global = nobs_nd.get_data()[0];
+    {
+        ONEDAL_PROFILER_TASK(allreduce_rows_count_global);
+        comm_.allreduce(rows_count_global, spmd::reduce_op::sum).wait();
+    }
+    if (res_op.test(result_options::min)) {
+        ONEDAL_ASSERT(input.get_partial_min().get_column_count() == column_count);
+        const auto min =
+            pr::table2ndarray_1d<Float>(q, input.get_partial_min(), sycl::usm::alloc::device);
+
+        { comm_.allreduce(min.flatten(q, {}), spmd::reduce_op::min).wait(); }
+        res.set_min(homogen_table::wrap(min.flatten(q, {}), 1, column_count));
+    }
+
+    if (res_op.test(result_options::max)) {
+        ONEDAL_ASSERT(input.get_partial_max().get_column_count() == column_count);
+        const auto max =
+            pr::table2ndarray_1d<Float>(q, input.get_partial_max(), sycl::usm::alloc::device);
+
+        { comm_.allreduce(max.flatten(q, {}), spmd::reduce_op::max).wait(); }
+        res.set_max(homogen_table::wrap(max.flatten(q, {}), 1, column_count));
+    }
+
+    if (res_op_partial.test(result_options::sum)) {
+        const auto sums_nd =
+            pr::table2ndarray_1d<Float>(q, input.get_partial_sum(), sycl::usm::alloc::device);
+        {
+            ONEDAL_PROFILER_TASK(allreduce_sums, q);
+            comm_.allreduce(sums_nd.flatten(q, {}), spmd::reduce_op::sum).wait();
+        }
+        const auto sums2_nd = pr::table2ndarray_1d<Float>(q,
+                                                          input.get_partial_sum_squares(),
+                                                          sycl::usm::alloc::device);
+        {
+            ONEDAL_PROFILER_TASK(allreduce_sums, q);
+            comm_.allreduce(sums2_nd.flatten(q, {}), spmd::reduce_op::sum).wait();
+        }
+        const auto sums2cent_nd =
+            pr::table2ndarray_1d<Float>(q,
+                                        input.get_partial_sum_squares_centered(),
+                                        sycl::usm::alloc::device);
+        {
+            ONEDAL_PROFILER_TASK(allreduce_sums, q);
+            comm_.allreduce(sums2cent_nd.flatten(q, {}), spmd::reduce_op::sum).wait();
+        }
```

Comment:
I think this piece is incorrect.
Centered squared sums, i.e. $s_{cent, j} = \sum_{i = 1, .., n} (x_{ij} - \mu_j)^2$  cannot be computed correctly until mean values $\mu_j$ are computed for the whole dataset. And means are computed later, in 'compute_all_metrics'.

Reply:
Yes, I recompute it here https://github.com/oneapi-src/oneDAL/pull/2693/files#diff-779be883bb3f6b3638c41cbb4c36f1d94ec3c1fbf9a4fdf12599ff71da19792bR73.
Does it make sense to remove Centered squared sums from partial results? And compute it only on finalize step?

Reply:
Has been discussed offline